### PR TITLE
Fix step shifting when accumulate gradient

### LINKED ISSUE #33671: Step shifting using total_batched_samples for gradient_accumulation_steps counting
### System Info

- `transformers` version: 4.39.0
- Platform: Linux-5.4.239-1.el7.elrepo.x86_64-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.24.6
- Safetensors version: 0.4.1
- Accelerate version: 0.28.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.2.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Issue Analysis
The `on_step_begin` callback is invoked when the `step` is divisible by `args.gradient_accumulation_steps` (i.e.,` step % args.gradient_accumulation_steps == 0`).
However, the `on_step_end` callback behaves differently. Its condition is as follows:
`total_batched_samples % args.gradient_accumulation_steps == 0 or is_last_step_and_steps_less_than_grad_acc`

Here, the `on_step_end` callback is triggered when `total_batched_samples` is divisible by `args.gradient_accumulation_steps`. It’s important to note that `step` is reset at the beginning of each epoch, whereas `total_batched_samples` is initialized to 0 at the start of training and persists across all epochs until training ends. 

### Expected Behavior:

When `gradient_accumulation_steps = N`, there should be exactly N sub-steps between the `on_step_begin` and `on_step_end` callbacks. This ensures that gradients are accumulated correctly before an optimization step occurs. The only exception to this rule is the last step in an epoch or the training run, where fewer sub-steps might exist.

### Problematic Behavior Example
The issue arises when `total_batched_samples` is not divisible by args.gradient_accumulation_steps. For example, if `steps_per_epoch = 3` and `gradient_accumulation_steps = 2`, we observe the following behavior:

Epoch 1:
* step 0: `on_step_begin` called, `on_step_end` not called (expected behavior)
* step 1: `on_step_begin` not called, `on_step_end` called (expected behavior)
* step 2: `on_step_begin` called, `on_step_end` not called (expected behavior)

Epoch 2:
* step 3: `on_step_begin` called (0 % 2 == 0), `on_step_end` called (4 % 2 == 0) (incorrect because on_step_end is called after only one sub step)
* step 4: `on_step_begin` not called (1 % 2 != 0), `on_step_end` not called (5 % 2 != 0) (incorrect)
* step 5: `on_step_begin` called (2 % 2 == 0), `on_step_end` called (6 % 2 == 0) (incorrect because on_step_end is called after only one sub step)

Epoch 3:
* step 6: `on_step_begin` called, `on_step_end` not called (expected behavior)
* step 7: `on_step_begin` not called, `on_step_end` called (expected behavior)
* step 8: `on_step_begin` called, `on_step_end` not called (expected behavior)

Note: `total_batched_samples` is incremented by 1 at the start of each step loop.

In this case, when the number of steps per epoch is not divisible by gradient_accumulation_steps, the callbacks only function correctly at intervals, leading to incorrect behavior during other epochs.

### PR DESCRIPTION
**Pull request by kibitzing**
# What does this PR do?

Fixes #33671

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr and @SunMarc

### CODE REVIEW

**Review by SunMarc — COMMENTED** [review]
Thanks for the detailed issue @kibitzing! Really nice job explaining what's happening! This looks like the right fix for gradient accumulation. In the past, we had the same behavior as this PR but without updating at the end of the epoch. See this [PR](https://github.com/huggingface/transformers/pull/22098/files) for more information. Since we do update after the end of each epoch, it make sense to not use `total_batched_samples` anymore. I see that in [accelerate](https://github.com/huggingface/accelerate/blob/4305033f8035defad0a87cd38e5c918e78510ba5/src/accelerate/accelerator.py#L1038), we only turn the gradient sync when we are at the end of the dataloader or `step+1 % args.gradient_accumulation_steps == 0`. Potentially, we can even remove the condition that we placed in transformers. 
From
```py
                if (
                    (step + 1) % args.gradient_accumulation_steps == 0
                    or
                    # last step in epoch but step is always smaller than gradient_accumulation_steps
                    is_last_step_and_steps_less_than_grad_acc
                ):
                    # the `or` condition of `is_last_step_and_steps_less_than_grad_acc` is not covered
                    # in accelerate. So, explicitly enable sync gradients to True in that case.
                    if is_last_step_and_steps_less_than_grad_acc:
                        self.accelerator.gradient_state._set_sync_gradients(True)
```
to simply
```python
if self.accelerator.sync_gradients:
```
Can you check if we have the same behavior ? cc @muellerz

**Comment by kibitzing** [comment]
Hello @SunMarc ,

Thank you for taking the time to review my PR! I appreciate your suggestion, it's a great idea and would indeed simplify the code.

However, after investigating and experimenting with this, I found that `self.gradient_state.end_of_dataloader` is always `False` in my case, preventing the step of `self.accelerator` from being reset. 
This happens because:
*  `self.gradient_state.end_of_dataloader` depends on `self.gradient_state.active_dataloader` 
    * `end_of_dataloader` [depends on](https://github.com/huggingface/accelerate/blob/4305033f8035defad0a87cd38e5c918e78510ba5/src/accelerate/state.py#L1197) `in_dataloader` 
    * `in_dataloader` [depends on](https://github.com/huggingface/accelerate/blob/4305033f8035defad0a87cd38e5c918e78510ba5/src/accelerate/state.py#L1250) `active_dataloader`
* and `active_dataloader` is generally [set to `None`](https://github.com/huggingface/accelerate/blob/4305033f8035defad0a87cd38e5c918e78510ba5/src/accelerate/state.py#L1163), unless explicitly configured otherwise.

While it's been interesting to dig deeper, I believe this issue might be outside the scope of the current PR. Hence, I suggest we stick with the previous approach for now.

Additionally, during my testing, I found a bug in the condition `is_last_step_and_steps_less_than_grad_acc`. It should use `or` instead of `and` for the expected behavior. If `and` is used, the last steps are always dropped if` args.gradient_accumulation_steps` is larger than `step_in_epoch`, which is highly likely to happen. I just removed the comparing `step_in_epoch` and `grad_accum_step` part because updating at the last step covers that case as well. I’ve fixed this in the PR as well because I believe it is directly related to gradient accumulation.

---

\+ It is not modified now but to use `self.accelerator.sync_gradients`, we should also remove https://github.com/huggingface/transformers/blob/f2c388e3f946862f657acc1e21b272ec946fc66c/src/transformers/trainer.py#L4820  because `_do_sync` function check this as well.

**Comment by kibitzing** [comment]
About failing test, I checked that it is not using Trainer which I fixed.
Is there any suggestion or comment about it?

**Comment by SunMarc** [comment]
After reading the code a bit more, I don't think we are performing a gradient update at the end of each epoch no ? I said that because as you saw the condition using `and` in `is_last_step_and_steps_less_than_grad_acc` + we have set grad_acc_kwargs["sync_with_dataloader"] = False. So, I feel like we are doing the right number of update across epoch. The only issue if that we indeed have an issue since there is no N sub-steps between the on_step_begin and on_step_end callbacks. 

```
Epoch 1

Steps 1, 2, 3, 4 (update parameter because (4 % 4 = 0)
Steps 5, 6, 7 (update because it's the last step)
Epoch 2

Step 8 (update parameter because 8 % 4 =0)
Steps 9, 10, 11, 12 (update parameter because 12 % 4 = 0)
Steps 13, 14 (update because it's the last step)
Epoch 3

Step 15, 16 (update parameter because 16 % 4 = 0)
Steps 17, 18, 19, 20 (update parameter because 16 % 4 = 0)
Steps 21 (update because it's the last step)
```

**Comment by kibitzing** [comment]
Thank you for reply @SunMarc, 
I may have misinterpreted the code initially. I agree that there's no issue with the update logic. I will go ahead and revert my changes to keep the condition `is_last_step_and_steps_less_than_grad_acc`.

> Thanks for exploring ! Could you share a minimal reproducer ? This might indeed require a fix on accelerate side.
Sure, I will!

**Comment by SunMarc** [comment]
> Thank you for reply @SunMarc,
> I may have misinterpreted the code initially. I agree that there's no issue with the update logic. I will go ahead and revert my changes to keep the condition is_last_step_and_steps_less_than_grad_acc.

I mean that the original issue you had (too many updates) might not exist

**Comment by kibitzing** [comment]
Yes, you are right, that issue does not exist. 
I thought it always update at last, but it doesn't because it is generally blocked by "steps_less_than_than_grad_acc" condition. 
Sorry for the confusion.

**Comment by SunMarc** [comment]
Not an issue ! I was confused also. Still, I will discuss with @muellerzr if it makes sense to switch to back to your idea with the update at each last step, just like how it is coded in accelerate. cc@mueller I'll keep you updated. 

Now, the issue is why self.gradient_state.end_of_dataloader doesn't work as expected. Feel free to open an issue on accelerate library with the reproducer ! Thanks a lot !

**Comment by kibitzing** [comment]
Okay, I will create a new issue regarding the `self.gradient_state.end_of_dataloader` and wait for updates on this PR. 
Thanks!

**Comment by kibitzing** [comment]
Hello @SunMarc, 

I revisited the issue I previously reported regarding `accelerate` and, embarrassingly, it turned out to be a problem with my own codebase. I was using `accelerate` together with a custom `Trainer` but had overridden `get_train_dataloader` without calling `accelerate.prepare(dataloader)`.

To explain the flow in more detail: 
1. `active_dataloader` is set to `None` at first. 
2. When `prepare(dataloader)` is called, it goes into `prepare_data_loader`, where it returns either a `DataLoaderShard` or `DataLoaderShard` that inherits from `DataLoaderStateMixin` as a new dataloader. 
3. These prepared dataloaders add an `active_dataloader` at every `begin()`. 
* Without calling `prepare`, it is clear that the `GradientState` doesn't have an active dataloader

Therefore, **there is no issue with the `accelerate` code itself**. Everything works correctly when `prepare` is called, and I confirmed that at the last batch, `self.gradient_state.end_of_dataloader=True` is set as expected. I also checked this with the pytorch `run_glue.py` example.

Thus, as you suggested, using `if self.accelerator.sync_gradients:` is perfectly fine and results in simpler code.

I’m going to change the current condition to this one and push the update.

**Comment by kibitzing** [comment]
Hello, I have updated the if condition as per @SunMarc's suggestion.
Do we have any updates regarding the update at each last step ?

**Comment by kibitzing** [comment]
@SunMarc 
Is there any update on this PR? 
If any additional work is needed, please let me know!

**Review by muellerzr — APPROVED** [review]
Nice! I feel this is a great simplification all around :)

**Comment by muellerzr** [comment]
(Last bit is resolving the conflicts)

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2372,21 +2369,7 @@ def _inner_training_loop(
...
-                    total_batched_samples % args.gradient_accumulation_steps == 0
-                    or
-                    # last step in epoch but step is always smaller than gradient_accumulation_steps
-                    is_last_step_and_steps_less_than_grad_acc
-                ):
-                    # the `or` condition of `is_last_step_and_steps_less_than_grad_acc` is not covered
-                    # in accelerate. So, explicitly enable sync gradients to True in that case.
-                    if is_last_step_and_steps_less_than_grad_acc:
-                        self.accelerator.gradient_state._set_sync_gradients(True)
-
```
Comment (SunMarc) [inline src/transformers/trainer.py:2389]: We decided to perform the grad acc at the end of the dataloader but in trainer, we will set self.accelerator.gradient_state._set_sync_gradients by ourselves and not rely on the values set by accelerate.accumulate

**Review by SunMarc — COMMENTED** [review]
Can you fix the merge comflits. We did a lot of [modification](https://huggingface.co/blog/gradient_accumulation) wrt to grad acc and it makes sense to have this now ! Feel free to ask us any question you have !

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -4786,8 +4769,6 @@ def create_accelerator_and_postprocess(self):
             # take the gradient_accumulation_steps setting from TrainingArguments.
             grad_acc_kwargs["num_steps"] = self.args.gradient_accumulation_steps
 
-        grad_acc_kwargs["sync_with_dataloader"] = False
```
Comment (kibitzing) [inline src/transformers/trainer.py:4789]: If I understand correctly, since we're managing it ourselves in the trainer, would it be safer to keep it set to False?

File: src/transformers/trainer.py (hunk 1 above)
Comment (kibitzing) [reply src/transformers/trainer.py:2389]: Okay, then I'll stick with the `do_sync` in the main branch code, and just replace the `total_batched_samples` with `step`.

File: src/transformers/trainer.py (hunk 2 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: Yeah, let's keep it set to False !

File: src/transformers/trainer.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: that's right !

**Comment by kibitzing** [comment]
Hello, @SunMarc @muellerzr 

Here is a summary of the new commits:

1. I merged the main branch and resolved conflicts.
2. I replaced `total_batched_samples` with `(step + 1)` and `removed steps_in_epoch <= args.gradient_accumulation_steps` condition to simplify, as we discussed before
> I see that in [accelerate](https://github.com/huggingface/accelerate/blob/4305033f8035defad0a87cd38e5c918e78510ba5/src/accelerate/accelerator.py#L1038), we only turn the gradient sync when we are at the end of the dataloader or step+1 % args.gradient_accumulation_steps == 0. Potentially, we can even remove the condition that we placed in transformers.

4. I reverted the previous changes and set `grad_acc_kwargs["sync_with_dataloader"] = False` as SunMarc mentioned, 
> we will set self.accelerator.gradient_state._set_sync_gradients by ourselves and not rely on the values set by accelerate.accumulate

If there are any issues or additional modifications needed, please let me know!

**Review by SunMarc — APPROVED** [review]
A lot better ! Thanks for your patience @kibitzing. If you are happy with the changes, feel free to merge the PR @muellerzr

**Comment by muellerzr** [comment]
failing tests come from main, we are in limbo 🫠

**Comment by muellerzr** [comment]
cc @ydshieh

**Comment by kibitzing** [comment]
Hello @muellerzr, I merged the main branch and it looks like it passed the tests! 😄

**Review by LysandreJik — APPROVED** [review]
Nice!
